doc/paper/paper.tex

```diff
@@ -0,0 +1,1012 @@
+    encoding format of the log’s messages.
+
+    A \textbf{channel} is a logical stream of messages within a log file. A channel is
+    associated with a schema, as well as a topic.
+
+    A \textbf{topic} is a name for a logical data stream, for instance \q{/images}. Each
+    channel has exactly one topic, multiple channels may have the same topic, and
+    channels with the same topic may have different schemas (unusual in a single
+    recording but common over time as an organization’s schemas evolve).
+
+    A \textbf{message} associates a channel with an array of bytes, which encodes a
+    message in the format described by the schema of the channel.
+
+    \subsection{As-of joins}
+    A central aspect of multimodal log search workloads is a similar operation to
+    what some SQL dialects call an \q{as-of join}. The idea of an as-of join is
+    to join records in one table with those in another based on a sequential or
+    temporal relationship - for instance, join records in one table with the
+    immediately preceding or succeeding records in another table, based on a
+    timestamp column.
+
+    As a result of the message-oriented architecture described in the previous
+    section, these kinds of queries are common in robotics. For an example query,
+    consider working on a self-driving car and needing to locate recent instances
+    where the car was taking a left hand turn in the rain, with more than 10
+    pedestrians present in an intersection, in order to debug some misbehavior
+    that has been noticed. The log messages that would indicate the left turn,
+    the rain, and the number of tracked pedestrians will likely occur on separate
+    topics, at unaligned timestamps and different logging frequencies. This means
+    what you really need to understand is the cases where variously-defined
+    \q{adjacent} readings on these topics match your search criteria.
+
+    As-of joins are a nonstandard SQL feature, and are not supported in all
+    analytics databases. For example none of BigQuery \cite{bigquery}, Redshift
+    \cite{redshift}, or Azure Synapse \cite{synapse}, support them.
+    DuckDB \cite{duckdb}, Clickhouse \cite{clickhouse}, and Snowflake \cite{snowflake} all do.
+
+    Regardless of whether an engine has explicit support for as-of joins, there
+    are complexities with putting SQL in front of the user and telling them to
+    write complex N-way as-of joins. First, while the robotics developer may have
+    some basic SQL experience, they are not a SQL expert, and particularly if the
+    dialect does not support explicit as-of joins the queries can become
+    extremely complex. Although one workable option is to hire SQL experts to
+    interface between the database and the robotics developer, one of dp3's
+    experimental goals is to avoid this solution and attempt instead to empower
+    the robotics developer directly. Second, analytics databases are not
```

Comment:
If this is one of the motivations of dp3, can we indicate this somewhere previously? This project has lots of potential, but I feel that the motivations are a bit scattered in the paper